x509util/crlcheck/crlcheck.go

```diff
@@ -19,67 +19,203 @@ package main
+	caFile      = flag.String("ca", "", "CA certificate file")
+	strict      = flag.Bool("strict", false, "Strict validation of CRL contents")
+	expectCerts = flag.Bool("cert", false, "Input files are certificates not CRLs")
+)
 
 func main() {
 	flag.Parse()
+
+	// Build a list of possible CA certs from command line arguments.
 	var caCerts []*x509.Certificate
 	if *caFile != "" {
 		caDataList, err := x509util.ReadPossiblePEMFile(*caFile, "CERTIFICATE")
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "%s: Failed to read CA cert data: %v\n", *caFile, err)
+			glog.Exitf("%s: Failed to read CA cert data: %v", *caFile, err)
 		}
 		for _, caData := range caDataList {
 			certs, err := x509.ParseCertificates(caData)
 			if err != nil {
-				fmt.Fprintf(os.Stderr, "%s: %v\n", *caFile, err)
+				glog.Errorf("%s: %v", *caFile, err)
 			}
-			if len(caCerts) == 0 {
-				fmt.Fprintf(os.Stderr, "%s: no certificates found\n", *caFile)
+			if len(certs) == 0 {
+				glog.Errorf("%s: no certificates found", *caFile)
 			}
 			caCerts = append(caCerts, certs[0])
 		}
 	}
 
-	errcount := 0
-	for _, filename := range flag.Args() {
-		dataList, err := x509util.ReadPossiblePEMURL(filename, "X509 CRL")
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "%s: failed to read data: %v\n", filename, err)
-			errcount++
-			continue
+	errored := false
+	for _, arg := range flag.Args() {
+		if *expectCerts {
+			if err := processCertArg(arg, caCerts); err != nil {
+				glog.Errorf("%s: failed to read certificate data: %v", arg, err)
+				errored = true
+			}
+		} else {
+			if err := processCRLArg(arg, caCerts); err != nil {
+				glog.Errorf("%s: failed to read CRL data: %v", arg, err)
+				errored = true
+			}
+		}
+	}
+
+	if errored {
+		os.Exit(1)
+	}
+}
+
+func processCRLArg(arg string, caCerts []*x509.Certificate) error {
+	dataList, err := x509util.ReadPossiblePEMURL(arg, "X509 CRL")
+	if err != nil {
+		return err
+	}
+	for _, data := range dataList {
+		if _, err := processCRL(data, caCerts); err != nil {
+			return err
 		}
+	}
+	return nil
+}
 
-		var crls []*x509.CertificateList
-		for _, data := range dataList {
-			certList, err := x509.ParseCertificateListDER(data)
+func processCRL(data []byte, caCerts []*x509.Certificate) (*x509.CertificateList, error) {
+	certList, err := x509.ParseCertificateListDER(data)
+	if certList == nil {
+		return nil, fmt.Errorf("CRL parse error: %v", err)
+	}
+	if err != nil && *strict {
+		return nil, fmt.Errorf("strict CRL parse error: %v", err)
+	}
+	glog.Infof("processing CRL:\n%s", x509util.CRLToString(certList))
+
+	verified := false
+	if len(caCerts) == 0 {
+		glog.Warningf("skipping signature validation as no CA certs available")
+		verified = true
+	}
+	var verifyErr error
+	for _, caCert := range caCerts {
+		if err := caCert.CheckCertificateListSignature(certList); err != nil {
+			verifyErr = err
+		} else {
+			glog.Infof("CRL signature verified against CA cert %q", x509util.NameToString(caCert.Subject))
+			verified = true
```

Comment:
Could potentially set `verifyErr` to nil here, just to avoid any potential for confusion?

Reply:
Done.